Extend preservable op list to custom ops and delete _preserve_ops.

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #136888
* #136887
* __->__ #136884

In this PR, we remove _preserve_ops API and solely use decomp_table moving forward. We use following mechanism to load up all custom ops.

1. We query whole C++ dispatcher at startup to materialize all possible CIA ops (aten + custom ops)
2. We walk through user exported program to find newly create custom ops that were not detected during the first round. We can't repeatedly call the first step because it is costly. 

Some notable changes in this PR:
1. Rename torch.export.core_aten_decomp to torch.export.core_op_decomp now it contains custom ops 
2. Moved some CIA related util funcvtions from torch/_decomp to torch/export. 
3. Delete _preserve_ops 
4. Leave torch._decomp.core_aten as it is because now we have export version of this table. 

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

tugsbayasgalan committed:
Extend preservable op list to custom ops

User can preserve custom ops as well. We do so by extending current logic to also query CIA ops from all possible op namespaces.

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63490973).

tugsbayasgalan committed:
Update on "Extend preservable op list to custom ops"


User can preserve custom ops as well. We do so by extending current logic to also query CIA ops from all possible op namespaces.

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Extend preservable op list to custom ops"


User can preserve custom ops as well. We do so by extending current logic to also query CIA ops from all possible op namespaces.

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63490973).

tugsbayasgalan committed:
Update on "Extend preservable op list to custom ops"


User can preserve custom ops as well. We do so by extending current logic to also query CIA ops from all possible op namespaces.

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63490973).

tugsbayasgalan committed:
Update on "Extend preservable op list to custom ops"


User can preserve custom ops as well. We do so by extending current logic to also query CIA ops from all possible op namespaces.

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63490973).

tugsbayasgalan committed:
Update on "Extend preservable op list to custom ops and delete _preserve_ops. "


In this PR, we remove _preserve_ops API and solely use decomp_table moving forward. We use following mechanism to load up all custom ops.

1. We query whole C++ dispatcher at startup to materialize all possible CIA ops (aten + custom ops)
2. We walk through user exported program to find newly create custom ops that were not detected during the first round. We can't repeatedly call the first step because it is costly. 

Some notable changes in this PR:
1. Rename torch.export.core_aten_decomp to torch.export.core_op_decomp now it contains custom ops 
2. Moved some CIA related util funcvtions from torch/_decomp to torch/export. 
3. Delete _preserve_ops 
4. Leave torch._decomp.core_aten as it is because now we have export version of this table. 

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

[ghstack-poisoned]

avikchaudhuri commented on torch/export/exported_program.py:
```diff
@@ -315,15 +441,32 @@ def _split_decomp_table_to_cia_and_python_decomp(
 
 
-def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
+def core_op_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
```
I think it's just better to name this `default_decompositions()`

bdhirsh reviewed (approved):
sgtm

tugsbayasgalan committed:
Update on "Extend preservable op list to custom ops and delete _preserve_ops. "


In this PR, we remove _preserve_ops API and solely use decomp_table moving forward. We use following mechanism to load up all custom ops.

1. We query whole C++ dispatcher at startup to materialize all possible CIA ops (aten + custom ops)
2. We walk through user exported program to find newly create custom ops that were not detected during the first round. We can't repeatedly call the first step because it is costly. 

Some notable changes in this PR:
1. Rename torch.export.core_aten_decomp to torch.export.core_op_decomp now it contains custom ops 
2. Moved some CIA related util funcvtions from torch/_decomp to torch/export. 
3. Delete _preserve_ops 
4. Leave torch._decomp.core_aten as it is because now we have export version of this table. 

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Extend preservable op list to custom ops and delete _preserve_ops. "


In this PR, we remove _preserve_ops API and solely use decomp_table moving forward. We use following mechanism to load up all custom ops.

1. We query whole C++ dispatcher at startup to materialize all possible CIA ops (aten + custom ops)
2. We walk through user exported program to find newly create custom ops that were not detected during the first round. We can't repeatedly call the first step because it is costly. 

Some notable changes in this PR:
1. Rename torch.export.core_aten_decomp to torch.export.core_op_decomp now it contains custom ops 
2. Moved some CIA related util funcvtions from torch/_decomp to torch/export. 
3. Delete _preserve_ops 
4. Leave torch._decomp.core_aten as it is because now we have export version of this table. 

Differential Revision: [D63490973](https://our.internmc.facebook.com/intern/diff/D63490973/)

[ghstack-poisoned]
